include/GafferScene/SceneAlgo.h

```diff
@@ -209,6 +209,12 @@ GAFFERSCENE_API SceneProcessor *objectTweaks( const ScenePlug *scene, const Scen
 /// Returns the last ShaderTweaks node to edit the specified attribute.
 GAFFERSCENE_API ShaderTweaks *shaderTweaks( const ScenePlug *scene, const ScenePlug::ScenePath &path, const IECore::InternedString &attributeName );
 
+/// Returns the name of a context variable in which the history methods store a unique value
+/// to disable the effects of the hash cache, so that the full upstream process can be examined.
+/// May be removed from a context to reenable the cache for expensive hash operations that are
+/// known to be irrelevant to the history.
```

Comment:
I'm trying to wrap my ahead when removing this var is permissible and/or useful.  Is the case where FilterResults was causing issues inside Set computations?  It seems like there is probably a more central place where we could remove this from the context ( ie. anywhere a filter is evaluated? ), but I guess that brings us back to how expensive creating contexts is, so we only want to remove it from the context in places where we know for sure an expensive calculation is happening.  Maybe some of this can wait until we do some in depth re-evaluation of Context creation as a whole, but one thing that would probably be fairly easy to do is some sort of fast short-circuit that doesn't create a new context if the variable you're trying to remove already isn't in the context?

Reply:
> I'm trying to wrap my ahead when removing this var is permissible and/or useful.

Permissible : when the upstream process will be discarded by `historyWalk()` anyway (implementation in SceneAlgo.cpp). So, anything which isn't the child of a ScenePlug, or a hop from one child to another (for instance a transform compute querying upstream attributes).

Useful : when hash is expensive, and is triggered by interactive use of `history()`.

> It seems like there is probably a more central place where we could remove this from the context

I think the ultimate solution is unlikely to be context-related at all. What `history()` is really trying to do is disable the effects of caching, so it can spy on the full upstream process tree using a Monitor. It may be that we make an API more explicitly suited for that purpose, and contrive some way of turning caching on/off at branch points in the tree, so the whole thing can be dealt with internal to the `history()` implementation.

I don't expect this optimisation to be all that widely applicable, but it is a massive benefit in this one specific circumstance. I didn't really add `historyIDContextName()` as an advertisement for wider usage, but just so I couldn't inadvertently break the optimisation by switching up the internal implementation in the future.